Added basic NER using prompt engineering

KTS-o7 opened the pull request:
Hello @willbakst !
I have tried some basic implementation on prompt engineering way of doing NER. There is a lot of improvements for structuring it . This is just a prototype from my side.
related to our discussion at #259 
- Provide me your feedback on how should this be improved ? What parts I should be rewriting etc.
- Looking forward for any other suggestions and guidance.  

Thanks in advance !

willbakst commented on docs/cookbook/named_entity_recognition.md:
```diff
@@ -0,0 +1,422 @@
+
+    Named Entity Recognition is a subtask of information extraction that seeks to locate and classify named entities in text into predefined categories such as person names, organizations, locations, etc. LLMs have revolutionized NER by enabling more context-aware and hierarchical entity recognition, going beyond traditional rule-based or statistical methods.
+
+## Setup and Imports
```
nit: take a look at the structure of the other cookbooks here. For setup, we should show installing the necessary packages and ensuring api keys are set.

we try to keep imports in the code cells where they are used to it's easier to grok what's going on

KTS-o7 replied:
Yes, I had written this in little hurry. Will fix it soon.

KTS-o7 replied:
Added structure and instructions

willbakst commented on docs/cookbook/named_entity_recognition.md:
````diff
@@ -0,0 +1,422 @@
+Now, let's create our main NER function using Groq's LLM:
+
+```python
+@groq.groq_call(
````
nit: just use `groq.call` here for enhanced readability

KTS-o7 replied:
Will be fixing it 👍🏼

KTS-o7 replied:
Fixed in new commit

willbakst commented on docs/cookbook/named_entity_recognition.md:
```diff
@@ -0,0 +1,422 @@
+
+## Testing with Various Scenarios
+
+To ensure robustness, it's crucial to test the NER system with diverse scenarios. Here's a function to run multiple test cases:
```
also, in the vein of closing #259 we should maybe show testing various levels ranging from 1 to 3?

KTS-o7 replied:
Yes i do want to implement examples showing various examples 1 to 3 or a little more.

KTS-o7 commented:
I have added the changes you suggested and put the files in right places in py and md formats.
fixed the following
- groq.call used instead of groq.groq_call
- used textwrap to indent
- made the test cases changes
- added simple ner and improved ner for changes.

KTS-o7 commented on docs/cookbook/named_entity_recognition.md:
```diff
@@ -0,0 +1,422 @@
+
+class NamedEntityRecognition(BaseModel):
+    entities: List[NamedEntity] = Field(description="List of top-level named entities found in the text")
+    given_text: str = Field(description="The text that was checked for named entities")
```
Dropped in the new commit

KTS-o7 commented on docs/cookbook/named_entity_recognition.md:
```diff
@@ -0,0 +1,422 @@
+    response_model=NamedEntityRecognition,
+    json_mode=True,
+)
+@prompt_template(
```
Tried adding something in the new commit.

KTS-o7 commented on docs/cookbook/named_entity_recognition.md:
```diff
@@ -0,0 +1,422 @@
+4. Developing a post-processing step to refine and validate the LLM's output.
+5. Exploring ways to optimize performance for real-time applications.
+
+By iterating on this foundation, you can create a powerful NER system tailored to your specific needs.
```
Changes added

KTS-o7 commented on docs/cookbook/named_entity_recognition.md:
```diff
@@ -0,0 +1,422 @@
+
+By iterating on this foundation, you can create a powerful NER system tailored to your specific needs.
+
+### Full Code Example
```
done

KTS-o7 commented on docs/cookbook/named_entity_recognition.md:
```diff
@@ -0,0 +1,422 @@
+
+The `test_cases` list contains various scenarios covering different domains and entity complexities, such as business and technology, politics and international relations, sports and entertainment, science and academia, complex nested entities, and historical and cultural references.
+
+## Conclusion and Further Improvements
```
changes made !

KTS-o7 commented on docs/cookbook/named_entity_recognition.md:
````diff
@@ -0,0 +1,422 @@
+test_ner_system(recognize_entities)
+```
+
+The `test_cases` list contains various scenarios covering different domains and entity complexities, such as business and technology, politics and international relations, sports and entertainment, science and academia, complex nested entities, and historical and cultural references.
````
Defined and shown in new commit !

KTS-o7 commented on docs/cookbook/named_entity_recognition.md:
````diff
@@ -0,0 +1,422 @@
+```python
+def print_entity_tree(entity, level=0):
+    indent = "  " * level
+    print(f"{indent}Entity: {entity.entity}, Label: {entity.label}, Parent: {entity.parent}")
````
added

KTS-o7 commented on docs/cookbook/named_entity_recognition.md:
```diff
@@ -0,0 +1,422 @@
+for entity in response.entities:
+    print_entity_tree(entity)
+print("\nGiven Text:")
+print(response.given_text)
```
added outputs at right places. followed the norm from other docs !

willbakst commented:
> I have added the changes you suggested and put the files in right places in py and md formats. fixed the following
> 
> * groq.call used instead of groq.groq_call
> * used textwrap to indent
> * made the test cases changes
> * added simple ner and improved ner for changes.

I'm not seeing your commit, make sure to push it a re-request my review when it's ready for another look :)

KTS-o7 commented:
I dont know why my commits arent showing up. I will close this PR and open a fresh one again. ( thats what stackoverflow workaround said tho :( )

KTS-o7 commented:
I think now it's visible for review @willbakst

willbakst commented on docs/cookbook/named_entity_recognition.md:
````diff
@@ -0,0 +1,269 @@
+First, ensure you have the necessary packages installed and API keys set:
+
+```bash
+pip install mirascope
````
nit: should be `pip install "mirascope[groq]"`

KTS-o7 replied:
Changed it !

willbakst commented on docs/cookbook/named_entity_recognition.md:
```diff
@@ -0,0 +1,269 @@
+    entity: str = Field(description="The entity found in the text")
+    label: str = Field(description="The label of the entity (e.g., PERSON, ORGANIZATION, LOCATION)")
+
+class SimpleNER(BaseModel):
```
we can drop this and just directly extract `list[SimplyEntity]`

KTS-o7 replied:
To keep the readability, i have followed this format. So that when you use the model in calls like 
```python
@groq.call(
    model="llama-3.1-8b-instant",
    response_model=SimpleNER,
    json_mode=True,
    call_params=groq.GroqCallParams(temperature=0.0),
)
```
In the response_model parameter it will be easy to look at, but can be changed.

KTS-o7 replied:
I have added the changes as requested

willbakst commented on docs/cookbook/named_entity_recognition.md:
```diff
@@ -0,0 +1,269 @@
+    parent: Optional[str] = Field(description="The parent entity if this entity is nested within another entity", default=None)
+    children: List['NestedEntity'] = Field(default_factory=list, description="Nested entities within this entity")
+
+class ImprovedNER(BaseModel):
```
same thing here, can just use `list[NestedEntity]`

KTS-o7 replied:
Again i thought it would be a better representation to keep things clear.

willbakst commented on docs/cookbook/named_entity_recognition.md:
````diff
@@ -0,0 +1,269 @@
+To ensure robustness, it's crucial to test the NER system with diverse scenarios. Here's a function to run multiple test cases:
+
+```python
+def test_ner_system(recognize_entities_func):
````
this test should actually test the input with an expected output rather than just printing things out

KTS-o7 replied:
I didnt understand how to validate it. Any particular way to validate them ?

willbakst replied:
the goal would be to have input text with a known output NER structure, run the NER against the input and then confirm the output matches the expected output.

willbakst commented on docs/cookbook/named_entity_recognition.md:
```diff
@@ -0,0 +1,269 @@
+
+## Background
+
+Named Entity Recognition is a subtask of information extraction that seeks to locate and classify named entities in text into predefined categories such as person names, organizations, locations, etc. LLMs have revolutionized NER by enabling more context-aware and hierarchical entity recognition, going beyond traditional rule-based or statistical methods.
```
we should add a note here that there are systems built specifically for NER that are much smaller / cheaper and can often get better results for the right tasks where using an LLM should generally be reserved for quick and dirty prototyping or tasks that require a more nuanced, open ended language based approach (for example, an NER system that accepts user input to guide the system might be easier to prototype and build with LLMs than a traditional NER system)

willbakst commented on examples/cookbook/named_entity_recognition.py:
note: make sure to update this code in line with comments from markdown

KTS-o7 replied:
Got it, will be adding comments !

KTS-o7 commented:
@willbakst ,
I have made changes to the usage of pydantic models. But the validation of NER test cases is something that i didnt get any ideas for.

Would be really helpful if you guys can finish the final part and merge :)
Looking forward to contribute more once I am done with semester finals.

willbakst commented:
> Would be really helpful if you guys can finish the final part and merge :)

Will do! We're a bit swamped this week, but I will aim to finish this up in time for the upcoming release so your contribution can be included :)
